dspace-server-webapp/src/main/java/org/dspace/app/rest/security/DSpaceCsrfAuthenticationStrategy.java

```diff
@@ -0,0 +1,83 @@
+/**
+ * The contents of this file are subject to the license and copyright
+ * detailed in the LICENSE and NOTICE files at the root of the source
+ * tree and available online at
+ *
+ * http://www.dspace.org/license/
+ */
+package org.dspace.app.rest.security;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.springframework.security.core.Authentication;
+import org.springframework.security.web.authentication.session.SessionAuthenticationException;
+import org.springframework.security.web.authentication.session.SessionAuthenticationStrategy;
+import org.springframework.security.web.csrf.CsrfToken;
+import org.springframework.security.web.csrf.CsrfTokenRepository;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+/**
+ * Custom SessionAuthenticationStrategy to be used alongside DSpaceCsrfTokenRepository.
+ * <P>
+ * Because DSpace is Stateless, this class only resets the CSRF Token if the client has attempted to use it (either
+ * successfully or unsuccessfully). This ensures that the Token is not changed on every request (since we are stateless
+ * every request creates a new Authentication object).
+ * <P>
+ * Based on Spring Security's CsrfAuthenticationStrategy:
+ * https://github.com/spring-projects/spring-security/blob/5.2.x/web/src/main/java/org/springframework/security/web/csrf/CsrfAuthenticationStrategy.java
+ */
+public class DSpaceCsrfAuthenticationStrategy implements SessionAuthenticationStrategy {
```

Comment:
I don't see any real flow in the proposed solution but if we need to identify one area where we could have hidden problem it is probable the use of a cookie for the csrf check. I'm not sure if having such cookie in place could for instance lead to false positive report by security tool. I'm just curious to know if you have investigated the use of a CSRF JWT based repository see for instance
https://stormpath.com/blog/csrf-protection-jwt-spring-security
https://github.com/spring-projects/spring-security/issues/5300

Reply:
@abollini : Just to clarify, yes, I had reviewed both those approaches previously, and found they didn't quite work as expected in a Stateless app where the client might be on a different *domain* than the server (i.e. cross domain requests).  

To try to explain it simply:
* I could have used a JWT for the CSRF token, but it's not recommended to use the Auth token as the CSRF token, so we'd need to generate a separate JWT (plus, unauthenticated users are given a CSRF Token to validate their login, whereas they don't get a JWT until after authenticating).  Rather than doing something much more custom, I went with the approach suggested by [OWASP of "Check if your framework has built-in CSRF protection & use it"](https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html).  So, in this PR, I'm letting Spring Security generate the CSRF token for us (it does so by generating a random UUID)
* While in this PR, the CSRF token *does* get stored in a Cookie, this is also the default behavior of [Spring Security's CSRF Protection](https://docs.spring.io/spring-security/site/docs/5.0.x/reference/html/csrf.html). 
    * By default, Spring Security *sends* the CSRF token in a Cookie (named XSRF-TOKEN) to the client, and expects the token to be sent back *both* in the XSRF-TOKEN cookie *and* in a X-XSRF-TOKEN Http Header. So, by default, Spring security verifies the Cookie & Header match.  This is *stateless*, but it doesn't work cross-domain as the client won't be able to read the XSRF-TOKEN cookie in order to send it back in a header.
    * My approach is similarly stateless, but gets around the cross-domain issue by sending the CSRF token to the client *both* in a cookie (DSPACE-XSRF-COOKIE) and in an Http Header (DSPACE-XSRF-TOKEN).  This ensures that a client on a different domain can read the token from the Http Header and send it back in a different Http Header (X-XSRF-TOKEN).  Then, Spring Security can still do the same validation -- comparing the value in the cookie (in this case DSPACE-XSRF-COOKIE) to that in the X-XSRF-TOKEN header.